route-rs-runtime/src/element/utils/icmp.rs

```diff
@@ -0,0 +1,262 @@
+    /// error packet
+    pub fn ttl_exceeded_error(
+        &self,
+        packet: &mut Ipv4Packet<Vec<u8>>,
+    ) -> Option<Ipv4Packet<Vec<u8>>> {
+        self.generic_error(
+            packet,
+            Icmpv4Message::TimeExceeded,
+            Icmpv4TimeExceeded::TtlExpired.into(),
+            &[0; 4],
+        )
+    }
+
+    // Common way of constructing and ICMPv4 error message, including the 8 byte ICMP header and
+    // as many bytes as possible of the offending IPv4 packet up to the configured MTU size
+    fn generic_error(
+        &self,
+        packet: &mut Ipv4Packet<Vec<u8>>,
+        msg_type: Icmpv4Message,
+        msg_code: u8,
+        _optional: &[u8; 4],
+    ) -> Option<Ipv4Packet<Vec<u8>>> {
+        if !self.should_generate_error(packet) {
+            return None;
+        }
+
+        assert!(
+            Icmpv4ErrorGenerator::is_error_type(msg_type),
+            "Provided ICMP message type must be an error type"
+        );
+
+        // TODO: this will need to vary if we add support for source routing
+        // The size of the IPv4 header plus the overhead for the ICMP header
+        let min_header_size = 20 + ICMP_HDR_LEN;
+
+        // How much of the offending packet will be copied into the ICMP error
+        let copied_payload_len = cmp::min(packet.total_len() as usize, self.mtu - min_header_size);
+
+        let error_pkt_repr = Ipv4Repr {
+            src_addr: self.src_ip,
+            dst_addr: packet.src_addr(),
+            protocol: IpProtocol::Icmp,
+            payload_len: copied_payload_len + ICMP_HDR_LEN,
+            hop_limit: self.ttl,
+        };
+
+        // Generate the layer 3 header
+        let buffer: Vec<u8> = vec![0x00; error_pkt_repr.buffer_len() + error_pkt_repr.payload_len];
+        let mut error_pkt = Ipv4Packet::new_unchecked(buffer);
+        error_pkt_repr.emit(&mut error_pkt, &ChecksumCapabilities::ignored());
+        error_pkt.set_dscp(packet.dscp());
+        error_pkt.fill_checksum();
+
+        // Generate the ICMP message
+        let mut icmp_error = Icmpv4Packet::new_unchecked(error_pkt.payload_mut());
+        icmp_error.set_msg_type(msg_type);
+        icmp_error.set_msg_code(msg_code);
+        // TODO: fill in bytes 5 - 8 with _optional data for ICMP errors that require it
+        icmp_error
+            .data_mut()
+            .clone_from_slice(&packet.clone().into_inner()[..copied_payload_len]);
+        icmp_error.fill_checksum();
+
+        Some(error_pkt)
+    }
+
+    // Performs checks based on RFC 1812 4.3.2.7 (When Not to Send ICMP Errors)
+    fn should_generate_error(&self, packet: &mut Ipv4Packet<Vec<u8>>) -> bool {
+        // Only the first fragment
+        if packet.frag_offset() != 0 {
+            return false;
+        }
+        // Avoid infinite loops, no errors from errors
+        if packet.protocol() == IpProtocol::Icmp {
+            if let Ok(icmp_packet) = Icmpv4Packet::new_checked(packet.payload_mut()) {
```

Comment:
packet.payload() does not work with underlying type Vec[u8] but packet.payload_mut does for some reason, so we need a mutable borrow even though we don't mutate anything?

Reply:
Yeah it is a weird implementation detail in smoltcp. I wonder if it is just something they never thought about?